pkg/lifecycle/lifecycle.go

```diff
@@ -79,171 +52,58 @@ func InitDependencies(rootFs billy.Filesystem, branchVersion string, currentChar
 	}
 
-	cycleLog(debug, "Getting branch version rules for: ", branchVersion)
-	// Initialize and check version rules for the current branch
-	dep.VR, err = GetVersionRules(branchVersion, debug)
+	// Initialize, load, and check version rules for the current branch
+	dep.VR, err = dep.rules(branchVersion, loadFromJSON)
 	if err != nil {
-		return nil, fmt.Errorf("encountered error while getting current branch version: %s", err)
+		return nil, err
 	}
 
-	// Get the filesystem and index.yaml path for the repository
-	dep.RootFs = rootFs
-
-	// Check if the assets folder and Helm index file exists in the repository
-	exists, err := filesystem.PathExists(dep.RootFs, path.RepositoryAssetsDir)
-	if err != nil {
-		return nil, fmt.Errorf("encountered error while checking if assets folder already exists in repository: %s", err)
-	}
-	if !exists {
-		return nil, fmt.Errorf("assets folder does not exist in the repository")
-	}
-	exists, err = filesystem.PathExists(dep.RootFs, path.RepositoryHelmIndexFile)
-	if err != nil {
-		return nil, fmt.Errorf("encountered error while checking if Helm index file already exists in repository: %s", err)
-	}
-	if !exists {
-		return nil, fmt.Errorf("Helm index file does not exist in the repository")
+	if err := checkFilePaths(dep.RootFs); err != nil {
+		return nil, err
 	}
 
 	// Get the absolute path of the Helm index file and assets versions map to apply rules
 	helmIndexPath := filesystem.GetAbsPath(dep.RootFs, path.RepositoryHelmIndexFile)
-	dep.assetsVersionsMap, err = getAssetsMapFromIndex(helmIndexPath, currentChart, debug)
-	if len(dep.assetsVersionsMap) == 0 {
-		return nil, fmt.Errorf("no assets found in the repository")
-	}
+	dep.AssetsVersionsMap, err = getAssetsMapFromIndex(helmIndexPath, currentChart)
 	if err != nil {
-		return nil, err // Abort and return error if the assets map is empty
+		return nil, err
+	}
+	if len(dep.AssetsVersionsMap) == 0 {
+		return nil, errorChartRepository
 	}
 
 	return dep, nil
 }
 
-// ApplyRules will populate all assets versions and paths, sort the versions,
-// and execute make remove for each chart and version.
-// After each chart removal, it will commit the changes in a single commit
-// for all versions of that chart.
-func (ld *Dependencies) ApplyRules(currentChart string, debug bool) error {
-	// Populate the assets versions and paths combining the index.yaml and assets folder
-	err := ld.populateAssetsVersionsPath(debug)
+func checkFilePaths(rootFs billy.Filesystem) error {
+	// Check if the assets folder and Helm index file exists in the repository
+	exists, err := filesystem.PathExists(rootFs, path.RepositoryAssetsDir)
 	if err != nil {
 		return err
 	}
-	// Sort the versions before removing
-	ld.sortAssetsVersions()
-
-	// Execute make remove for each chart and version that is not in the lifecycle.
-	// Commit after each chart removal.
-	removedAssetsVersions, err := ld.removeAssetsVersions(debug)
+	if !exists {
+		return errorChartRepository
```

Comment:
Is this error used only here? If so, you don't need to predeclare it above.

Reply:
There's a few other places that uses this error, and IMO we should declare every custom error, so when we unit test we have a variable to compare the expected error.

Reply:
The error is used in 3 different places for now. 
 @briandowns 
 
It is a general one that will be thrown if you have done something really nasty to the charts repository local folder. 
That is why I am not doing 3 distinct errors and only a general one. 
@rafaelbreno